Add 10 new lexical selection rules

albertonl opened the pull request:
- Rule 1: _much of the_ to be translated to _gran parte de_ instead of _mucho de_ ("fire destroyed much of the complex in 1512" was translated "el fuego destruyó mucho del complejo en 1512").

- Rule 2: _Middle Ages_ to be translated to _Edad Media_ instead of _Edades Medias_ ("[...] strategically important during the Middle Ages" was translated "[...] estratégicamente importante durante las Edades Medias").

- Rule 3: after a comma, _as [prpers/np] [was/were] [here/there]_ to be translated to _ya que [prpers/np] [estar (ppi)] [aquí/allí]_ instead of _como [prpers/np] [ser (prs)] [aquí/allí]_ ("[...] strategically important during the Middle Ages, as it was located [...]" was translated "[...] estratégicamente importante durante las Edades Medias (_see Rule 2_), como esté localizado [...]")

- Rule 4: _the [ordinal] to [verb]_ to be translated to _el [ordinal] en [verb]_ instead of _el [ordinal] para [verb]_ ("the first to include representatives" was translated "el primero para incluir representantes").

- Rule 5: _document storage_ to be translated to _almacenamiento de documentos_ instead of _almacenamiento de documento_.

- Rule 6: _as part of_ to be translated to _como parte de_ instead of _tan parte de_ ("[...] was demolished as part of this work" was translated to "estuvo derribado tan parte de esta obra").

- Rule 7: _alterations caused controversy_ to be translated to _[las] alteraciones causaron controversia_ instead of _alteraciones controversia causada_.

- Rule 8: _focused on ease_ to be translated to _enfocado en [la] facilidad_ instead of _enfocado sobre [la] facilidad_ (also with _difficulty_).

- Rule 9: _open-source_ to be translated to _código abierto_ instead of _fuente abierta_.

- Rule 10: _at the time_ to be translated to _en ese tiempo_ instead of _en el tiempo_.

mr-martian commented:
Some problems with rule 2:

The input reaches `lrx-proc` (which implements these rules) as `^middle<adj>/medio<adj>$ ^age<n><pl>/edad<n><f><pl>$`

The rule you have is
```xml
    <rule weight="1.0">
      <match lemma="Middle" tags="adj">
        <select lemma="Edad" tags="n.f.sg"/>
      </match>
      <match lemma="Ages" tags="n.*">
        <select lemma="Media" tags="adj.f.sg"/>
      </match>
    </rule>
```
What this rule says:
"Look for `^Middle<adj>/Edad<n><f><sg>/[other stuff]$ ^Ages<n><...>/Media<adj><f><sg>/[other stuff]$` and select the `Edad` reading of `Middle` and the `Media` reading of `Ages`."

The issues:

- This rule would only match if the lemmas of the input were capitalized (an aspect of the behavior of the `lrx-comp` that might be worth changing).
- Lexical selection can only select analyses that already exist and "Ages" won't have a singular reading available.
- Switching the order of the words as you do here is the job of structural transfer. After lexical selection all the words should be in exactly the same order they were in in English.

Similar problems show up in several of your other rules.

Two general suggestions:
1. Run the phrase you're trying to fix through `apertium eng-spa-biltrans` and remember that every `<match>` has to correspond to a thing on the left and every `<select>` and `<remove>` has to correspond to one of the things on the right.
2. Most of your rules would probably be better as transfer rules in the `.t1x` file and we can add structural transfer tasks if you'd like to work on that.

mr-martian commented:
I just finished checking the rest.

Rule 6 I think will actually work. Rule 7 gets wrong input from English disambiguation. All the rest have the problem of trying to add readings that didn't exist before.

albertonl commented:
I have now modified, deleted and added rules as you told me, both here and on IRC. This is the explanation of the current rules:

- Rule 1: previous rule 6.

- Rule 2: _the country's_ (and similar) _status_ to be translated to _el estatus del país_ instead of _el estado del país_ ("[...] confirmed the country's status as a global military power" was translated "[...] confirmaron el estado del país como poder militar global (see rule 3)").

- Rule 3: _global_ (and similar)_ military power_ to be translated to _potencia militar global_ instead of _poder militar global_ ("[...] confirmed the country's status as a global military power" was translated "[...] confirmaron el estado del país (see rule 2) como poder militar global").

- Rule 4: _socioeconomic_ (and similar) _performance_ to be translated to _rendimiento socioeconómico_ instead of _actuación socioeconómica_ ("very high in measures of socioeconomic performance" was translated "muy alto en medidas de actuación socioeconómica").

- Rule 5: _from a letter dated_ (+date) to be translated to _de una carta datada [de]_ instead of _de una letra datada [de]_ ("[...] is from a letter dated January 2, 1776" was translated "[...] es de una letra enero datado 2, 1776" (the translated date is also completely messed up)).

- Rule 6: _draft/version of the article_ to be translated to _borrador/versión del artículo_ instead of _borrador/versión de la prenda_ ("The second draft of the Articles of Confederation" was translated "El segundo borrador de las Prendas de Confederación").

- Rule 7: _use_ (and similar) _the term_ to be translated to _usar el término_ instead of _usar el plazo_ ("[...] it was written before or after Dickinson used the term in his June 17 draft" was translated "[...] esté escrito antes de que o después de *Dickinson utilizado el plazo en su junio 17 borrador" (there is also a lot of messed up stuff here, but not possible to fix with lexical selection rules)).

- Rule 8: _[personal pronoun/proper noun] appear(s)_ to be translated to _[personal pronoun/proper noun] aparece_ instead of _[personal pronoun/proper noun] parece_ ("it appears in the name "District of Columbia."" was translated "parece en el Distrito "de nombre de *Columbia."" (also with other stuff to be fixed)).

- Rule 9: _collection of independent states_ (and similar) to be translated to _grupo de estados independientes_ instead of _colección de estados independientes_ ("a description of a collection of independent states" was translated "una descripción de una colección de estados independientes").

- Rule 10: _over the year_ (and similar time expressions) to be translated to _durante el año_ instead of _sobre el año_ ("Over the years, [...]" was translated "Sobre los años, [...]").

Text used to look for translation mistakes extracted from: https://en.wikipedia.org/wiki/United_States

mlforcada commented on apertium-eng-spa.eng-spa.lrx:
```diff
@@ -8715,7 +8715,185 @@ convertirse en un personaje de moda o importante. -->
+    -->
+    <rule weight="1.0">
+      <match lemma="as" tags="*">
+        <select lemma="como" tags="cnjadv"/>
```
Is this the correct tag for "como" when it introduces a noun phrase?
Wouldn't it be como.pr?

albertonl replied:
I'm sorry, I don't completely understand. What do you exactly mean?

mlforcada replied:
When introducing a noun phrase, "como" is a preposition rather than an adverbial conjunction. 
However, this does not appear as a translation of "as" in the bilingual dictionary, as far as I know. Should it be added? Do we need to add items in "select" to the dictionaries? (I am not too familiar with lexical selection software in Apertium). It does not make sense before a noun phrase (it would be OK for a verb phrase, such as "as follows")

mr-martian replied:
English tags "as part of" as `^as<cnjadv>$ ^part<n><sg>$ ^of<pr>$` which then comes out of biltrans as `^as<cnjadv>/como<cnjadv>/cuando<cnjadv>$ ^part<n><sg>/parte<n><f><sg>/papel<n><f><sg>$ ^of<pr>/de<pr>$` at which point lexical selection must choose either `como<cnjadv>` or `cuando<cnjadv>`, so yes, changes would need to be made elsewhere to get a preposition here.

albertonl replied:
But, what preposition could we substitute it with?

mr-martian replied:
I think the answer to your question is that Mikel is saying it should be `^como<pr>$` rather than `^como<cnjadv>$`, but that isn't in the dictionary and the English tagger doesn't mark "as" as a preposition in this case.

mlforcada replied:
Exactly.

mlforcada commented on apertium-eng-spa.eng-spa.lrx:
```diff
@@ -8715,7 +8715,185 @@ convertirse en un personaje de moda o importante. -->
+    -->
+    <rule weight="1.0">
+      <or>
+        <match lemma="socioeconomic" tags="adj"/>
```
"socioeconomic performance"?

albertonl replied:
I found that in the Wikipedia page:

"States continues to rank very high in measures of socioeconomic performance, including average wage, median income, median wealth, human development, per capita GDP, and worker productivity."

mlforcada replied:
Good example.

mlforcada commented on apertium-eng-spa.eng-spa.lrx:
```diff
@@ -8715,7 +8715,185 @@ convertirse en un personaje de moda o importante. -->
+      spa: durante el año
+    -->
+    <rule weight="1.0">
+      <match lemma="over" tags="pr">
```
Isn't it rather "a lo largo de los años"?

albertonl replied:
In this case, `eng-spa-biltrans` only shows _sobre_ and _durante_ as possible translations of _over_.

That "a lo largo de los años" might be a thing if _year_ is in plural, something that I haven't noticed though. Maybe I could change it so that it only matches _year_ just when it's singular?

mlforcada replied:
Can't one add a multiword translation in the dictionary just to be selected by this rule?
Does the output of <select> really need to match the dictionary or can it be anything?

mr-martian replied:
Lexical selection must choose from among the outputs of the dictionary, but yes, multiwords are definitely an option that I forgot about when making my earlier suggestions.

albertonl replied:
Alright, so if you want, I can add that multiword translation.

mlforcada replied:
Approval won't depend on that, I guess.

mlforcada commented:
Some comments. I cannot compile it in this machine, though (I get warnings about the "ig" element). As regards "the Middle Ages" → "la Edad Media", I believe this is clearly something that could be taken care of in dictionaries, rather than during transfer, as it is a fixed expression.
Some more comments in the code itself.

albertonl commented:
Sorry, I tried to revert commit [b7d318a](https://github.com/apertium/apertium-eng-spa/pull/14/commits/b7d318a382324d7bca7141b549fad863e7f582ed), but it seems like I wrongly pushed. However, all modifications there get back to its original form in commit [8e32c4b](https://github.com/apertium/apertium-eng-spa/pull/14/commits/8e32c4bbf69a922fa102c6d27c61c512447708e2).

Well, as a summary of these last three commits, I added an _a lo largo de_ translation for _over_ (and _throughout_, since it might also be used in similar sentences), and I added a new rule, for the task. However, we can still be working on the _as part of_ one.

albertonl commented:
@mlforcada _como_ as a preposition is already in the bidix, but it hasn't been recognized as so in any of the example sentences I've used, though
